src/Model/Term.php

```diff
@@ -4,20 +4,145 @@
+			}
+		}
+
+		$data = ' ' . $message['title'] . ' ' . $message['body'] . ' ' . $tags_string . ' ';
+
+		// ignore anything in a code block
+		$data = preg_replace('/\[code\](.*?)\[\/code\]/sm', '', $data);
+
+		$tags = [];
+
+		$pattern = '/\W\#([^\[].*?)[\s\'".,:;\?!\[\]\/]/ism';
+		if (preg_match_all($pattern, $data, $matches)) {
+			foreach ($matches[1] as $match) {
+				$tags['#' . strtolower($match)] = '';
+			}
+		}
+
+		$pattern = '/\W([\#@])\[url\=(.*?)\](.*?)\[\/url\]/ism';
+		if (preg_match_all($pattern, $data, $matches, PREG_SET_ORDER)) {
+			foreach ($matches as $match) {
+				$tags[$match[1] . strtolower(trim($match[3], ',.:;[]/\"?!'))] = $match[2];
+			}
+		}
+
+		foreach ($tags as $tag => $link) {
+			if (substr(trim($tag), 0, 1) == '#') {
+				// try to ignore #039 or #1 or anything like that
+				if (ctype_digit(substr(trim($tag), 1))) {
+					continue;
+				}
+
+				// try to ignore html hex escapes, e.g. #x2317
+				if ((substr(trim($tag), 1, 1) == 'x' || substr(trim($tag), 1, 1) == 'X') && ctype_digit(substr(trim($tag), 2))) {
+					continue;
+				}
+
+				$type = TERM_HASHTAG;
+				$term = substr($tag, 1);
+			} elseif (substr(trim($tag), 0, 1) == '@') {
+				$type = TERM_MENTION;
+				$term = substr($tag, 1);
+			} else { // This shouldn't happen
+				$type = TERM_HASHTAG;
+				$term = $tag;
+			}
+
+			if ($message['uid'] == 0) {
+				$global = true;
+				dba::update('term', ['global' => true], ['otype' => TERM_OBJ_POST, 'guid' => $message['guid']]);
+			} else {
+				$global = dba::exists('term', ['uid' => 0, 'otype' => TERM_OBJ_POST, 'guid' => $message['guid']]);
+			}
+
+			dba::insert('term', [
+				'uid'      => $message['uid'],
+				'oid'      => $itemid,
+				'otype'    => TERM_OBJ_POST,
+				'type'     => $type,
+				'term'     => $term,
+				'url'      => $link,
+				'guid'     => $message['guid'],
+				'created'  => $message['created'],
+				'received' => $message['received'],
+				'global'   => $global
+			]);
+
+			// Search for mentions
+			if ((substr($tag, 0, 1) == '@') && (strpos($link, $profile_base_friendica) || strpos($link, $profile_base_diaspora))) {
+				$users = q("SELECT `uid` FROM `contact` WHERE self AND (`url` = '%s' OR `nurl` = '%s')", $link, $link);
+				foreach ($users AS $user) {
+					if ($user['uid'] == $message['uid']) {
+						dba::update('item', ['mention' => true], ['id' => $itemid]);
+						dba::update('thread', ['mention' => true], ['iid' => $message['parent']]);
+					}
+				}
+			}
+		}
+	}
+
+	public static function insertFromItemUri($itemuri, $uid)
+	{
+		$messages = dba::select('item', ['id'], ['uri' => $itemuri, 'uid' => $uid]);
+
+		if (DBM::is_result($messages)) {
+			while ($message = dba::fetch($messages)) {
+				self::insertFromItemId($message['id']);
+			}
+			dba::close($messages);
+		}
+	}
+
+
 	/**
 	 * @param integer $itemid item id
 	 * @return void
 	 */
-	public static function createFromItem($itemid)
+	public static function insertFromItemFileById($itemid)
```

Comment:
Why not ```insertFileFromItemId```?

BTW: Eventually I want to split the "term" table into one for every purpose. This should increase performance. But this won't be for this release.

Reply:
This function inserts terms from the `file` field of a given item. The more general purpose `insertFromItemId()` takes from other fields. By the way I feel like `insertFromItemFileById()` should be called from `insertFromItemId()`, but I didn't want to alter the current workings.

Reply:
This is a difficult naming, since it inserts new terms from a `file` field of an `item` identified by its `id`.

`insertFileFromItemId` sounds like we're inserting files, which isn't the current behavior. The "correct" name would be `insertFromFileOfItemById` but it's ugly as hell.

Reply:
I would like to keep it not so complicated :-)

Question is if we should split the files already in advance to the splitting of the database tables?

Reply:
I don't like ```insertFromFileFieldByItemId``` at all - we should use it :-)

Reply:
Done.